### DISCUSSION TITLE
Does try_map throw away the error?

### DISCUSSION THREAD

**Discussion by viridia** [winnow-rs/winnow#432] [Q&A]
I've been trying to figure out how to extract the error returned by `try_map`. I have a case where there are several possible errors (all in an enum), and I want to be able to print a different error message depending on which error occurred. However, when I look at the error result from the parser, there's no trace of the error that was returned from `try_map`.

I've looked at all the examples, and searched on the web (for hours), but all of the examples I have found don't appear to actually use the error result. There are several examples that show detecting a number conversion error, but they don't appear to actually use the error result from try_map, they just detect that an error happened.

**Comment by epage** [comment]
`try_map` takes the user-provided error and wraps it into the Parsers error type using `FromExternalError::from_external_error` ([source](https://docs.rs/winnow/latest/src/winnow/combinator/parser.rs.html#132))..

So the answer is dependent on your code  If you are using the default error type, `ContextError`, then the error gets boxed ([source](https://docs.rs/winnow/latest/src/winnow/error.rs.html#544-550)) and passed up the stack..

**Reply by viridia** [reply]
I'm using the default error type, that is I'm using PResult without overriding the error param.

When I print out the error, here's what I see:
```
error: ParseError {
    input: "&",
    offset: 0,
    inner: ContextError {
        context: [
            Label(
                "style property",
            ),
        ],
        cause: None,
    },
}
```
How do I access the boxed error? The doc for `ContextError` doesn't mention this.

Also BTW, the tutorial section on "error reporting" is somewhat misleading, in that while it illustrates how to generate errors, it says nothing about how to actually print them out or display them to the user.

**Reply by epage** [reply]
> How do I access the boxed error? The doc for ContextError doesn't mention this.

We have a [`ContextError::cause` method](https://docs.rs/winnow/latest/winnow/error/struct.ContextError.html#method.cause).  You would then need to dynamic cast it to your specific type.  The problem is that the `cause` is empty.  To understand why, I'd need a reproduction case.

> Also BTW, the tutorial section on "error reporting" is somewhat misleading, in that while it illustrates how to generate errors, it says nothing about how to actually print them out or display them to the user.

If you are in the happy path, like the `FromStr` examples, then you get a form of error printing for free.  What do to outside of that is so open ended, I'm not sure what to cover.  We do have #104 for alternative forms of reporting.

However, I'm not sure whats "misleading" about that vs just not covering the topic fully, especially for a beginning document.  For example, I expect #104 to be in the special topics and we could link out to that.

**Reply by viridia** [reply]
All right, I figured it out. The error wasn't happening where I thought it was :/
